client/network/src/service/tests.rs

```diff
@@ -272,3 +272,73 @@ fn notifications_state_consistent() {
 		}
 	});
 }
+
+#[test]
+fn lots_of_incoming_peers_works() {
+	let listen_addr = config::build_multiaddr![Memory(rand::random::<u64>())];
+
+	let (main_node, _) = build_test_full_node(config::NetworkConfiguration {
+		notifications_protocols: vec![(ENGINE_ID, From::from(&b"/foo"[..]))],
+		listen_addresses: vec![listen_addr.clone()],
+		in_peers: u32::max_value(),
+		transport: config::TransportConfig::MemoryOnly,
+		.. config::NetworkConfiguration::new_local()
+	});
+
+	let main_node_peer_id = main_node.local_peer_id().clone();
+
+	// We spawn background tasks and push them in this `Vec`. They will all be waited upon before
+	// this test ends.
+	let mut background_tasks_to_wait = Vec::new();
+
+	for _ in 0..256 {
```

Comment:
As I mentioned in another conversation already, the test often executes very slowly on my machine. Even if I change this line to `for _ in 0..3 {` I sometimes (but not always) get execution times like this:

```
> time cargo test lots_of_incoming_peers_works
    Finished test [unoptimized + debuginfo] target(s) in 0.28s
     Running /home/user/workspace/substrate/target/debug/deps/sc_network-53c6b71fe56316d8

running 1 test
test service::tests::lots_of_incoming_peers_works ... test service::tests::lots_of_incoming_peers_works has been running for over 60 seconds
test service::tests::lots_of_incoming_peers_works ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 42 filtered out


________________________________________________________
Executed in   67.74 secs   fish           external
   usr time  1298.09 millis  288.00 micros  1297.81 millis
   sys time  193.63 millis  147.00 micros  193.49 millis
```

It makes me wonder if some task is not properly woken up at some point.

Reply:
@twittner Just curious since I haven't tried running it at all, but which version of `async-std` are you running this with? I'm asking because the `dev-dependencies` of this crate state `async-std = "1.5"` which allows `1.6` which seems to have some issues (e.g. the libp2p noise smoke tests stalled completely with that version).

Reply:
`async-std` version 1.5.0, cf. [`Cargo.lock`](https://github.com/paritytech/substrate/blob/c40852882e9886156677775ea89cf498e9f69a36/Cargo.lock#L188).

Reply:
Do you think there's anything wrong in the test itself?
I don't see any problem and I'd rather suspect an issue in the networking code as a whole.
If you agree, I suggest to go on with this PR and open an issue?

Reply:
Yeah, the test is probably not to blame for this. Since CI seems to cope this is not a blocker.